crates/matrix-sdk-ui/tests/integration/sync_service.rs

```diff
@@ -0,0 +1,213 @@
+use std::{
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use matrix_sdk_test::async_test;
+use matrix_sdk_ui::sync_service::{SyncService, SyncServiceState};
+use serde_json::json;
+use wiremock::{Match as _, Mock, MockGuard, MockServer, Request, ResponseTemplate};
+
+use crate::{
+    logged_in_client,
+    sliding_sync::{PartialSlidingSyncRequest, SlidingSyncMatcher},
+};
+
+/// Sets up a sliding sync server that use different `pos` values for the
+/// encrptyion and the room sync. Will respond after 50 milliseconds.
+async fn setup_mocking_sliding_sync_server(server: &MockServer) -> MockGuard {
+    let encryption_pos = Mutex::new(0);
+    let room_pos = Mutex::new(0);
+
+    Mock::given(SlidingSyncMatcher)
+        .respond_with(move |request: &Request| {
+            let partial_request: PartialSlidingSyncRequest = request.body_json().unwrap();
+            // Repeat the transaction id in the response, to validate sticky parameters.
+            let mut pos = match partial_request.conn_id.as_deref() {
+                Some("encryption") => encryption_pos.lock().unwrap(),
+                Some("room-list") => room_pos.lock().unwrap(),
+                _ => panic!("unexpected conn id {:?}", partial_request.conn_id),
+            };
+
+            *pos += 1;
+            let pos_as_str = (*pos).to_string();
+
+            ResponseTemplate::new(200)
+                .set_body_json(json!({
+                    "txn_id": partial_request.txn_id,
+                    "pos": pos_as_str
+                }))
+                .set_delay(Duration::from_millis(50))
+        })
+        .mount_as_scoped(server)
+        .await
+}
+
+#[async_test]
+async fn test_sync_service_state() -> anyhow::Result<()> {
+    let (client, server) = logged_in_client().await;
+
+    let _guard = setup_mocking_sliding_sync_server(&server).await;
+
+    let sync_service =
+        SyncService::builder(client).with_encryption_sync(false, None).build().await.unwrap();
+
+    // At first, the sync service is sleeping.
+    assert_eq!(sync_service.state().get(), SyncServiceState::Idle);
+    assert!(server.received_requests().await.unwrap().is_empty());
+
+    // After starting, the sync service is, well, running.
+    sync_service.start().await?;
+    assert_eq!(sync_service.state().get(), SyncServiceState::Running);
+
+    // Restarting while started doesn't change the current state.
+    sync_service.start().await?;
+    assert_eq!(sync_service.state().get(), SyncServiceState::Running);
+
+    // Let the server respond a few times.
+    tokio::time::sleep(Duration::from_millis(300)).await;
+
+    // Pausing will stop both syncs, after a bit of delay.
+    sync_service.pause()?;
+
+    // Since this is racy: either it's already terminated, or we wait a bit and then
+    // it'll be terminated.
```

Comment:
Try running `yield_now().await` to see if it solves your flakyness here.

Reply:
No luck here.